File: functions/gocardless-sync-node/src/gocardless.js
/**
 * GoCardless API Client
 */

const BASE_URL = 'https://bankaccountdata.gocardless.com/api/v2';
const DEFAULT_TIMEOUT_MS = 20000;
const MAX_RETRIES = 3;

let accessToken = null;
let tokenExpiresAt = 0;

async function getAccessToken() {
  // Return cached token if still valid (with 30s buffer)
  if (accessToken && Date.now() / 1000 < (tokenExpiresAt - 30)) {
    return accessToken;
  }

  const url = `${BASE_URL}/token/new/`;
  const response = await fetch(url, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      secret_id: process.env.GOCARDLESS_SECRET_ID,
      secret_key: process.env.GOCARDLESS_SECRET_KEY,
    }),
  });

  if (!response.ok) {
    throw new Error(`Failed to get GoCardless token: ${response.status}`);
  }

  const data = await response.json();
  accessToken = data.access;
  tokenExpiresAt = Math.floor(Date.now() / 1000) + data.access_expires;

  return accessToken;
}

async function request(path, params) {
  const token = await getAccessToken();
  const url = new URL(`${BASE_URL}${path}`);
  
  if (params) {
    Object.entries(params).forEach(([key, value]) => {
      if (value) url.searchParams.set(key, value);
    });
  }

  for (let attempt = 0; attempt < MAX_RETRIES; attempt++) {
    try {
      const response = await fetch(url.toString(), {
        headers: { Authorization: `Bearer ${token}` },
        signal: AbortSignal.timeout(DEFAULT_TIMEOUT_MS),
      });

      if (!response.ok) {
        throw new Error(`GoCardless API error: ${response.status}`);
      }

      return await response.json();
    } catch (error) {
      if (attempt === MAX_RETRIES - 1) throw error;
      await new Promise(resolve => setTimeout(resolve, 1000 * (attempt + 1)));
    }
  }
}

export async function getTransactions(accountId, dateFrom) {
  const params = {};
  if (dateFrom) params.date_from = dateFrom;
  
  return request(`/accounts/${accountId}/transactions/`, params);
}

export async function getBalances(accountId) {
  return request(`/accounts/${accountId}/balances/`);
}

File: functions/gocardless-sync-mongo-python/src/gocardless.py
"""GoCardless API client helpers."""

import time

import requests

GOCARDLESS_BASE_URL = "https://bankaccountdata.gocardless.com/api/v2"
DEFAULT_TIMEOUT = 20
MAX_RETRIES = 3


class GoCardlessClient:
    def __init__(self, secret_id, secret_key):
        self.secret_id = secret_id
        self.secret_key = secret_key
        self._access_token = None
        self._token_expires_at = 0

    def _get_access_token(self):
        if self._access_token and time.time() < (self._token_expires_at - 30):
            return self._access_token
        
        url = f"{GOCARDLESS_BASE_URL}/token/new/"
        payload = {"secret_id": self.secret_id, "secret_key": self.secret_key}
        try:
            response = requests.post(url, json=payload, timeout=DEFAULT_TIMEOUT)
            response.raise_for_status()
            
        except Exception as e:
            raise e

        token_data = response.json()
        self._access_token = token_data["access"]
        self._token_expires_at = time.time() + token_data["access_expires"]
        return self._access_token

    def _request(self, path, params=None):
        
        token = self._get_access_token()
        headers = {"Authorization": f"Bearer {token}"}
        url = f"{GOCARDLESS_BASE_URL}{path}"

        for attempt in range(MAX_RETRIES):
            try:
                response = requests.get(
                    url, headers=headers, params=params, timeout=DEFAULT_TIMEOUT
                )
                response.raise_for_status()
                return response.json()
            except Exception as e:
                if attempt == MAX_RETRIES - 1:
                    raise e
                time.sleep(1)

    def get_transactions(self, account_id, date_from=None):
        params = {"date_from": date_from} if date_from else None
        return self._request(f"/accounts/{account_id}/transactions/", params)

    def get_balances(self, account_id):
        return self._request(f"/accounts/{account_id}/balances/")



